---
title: "GWAS Analysis"
output: html_notebook
---


```{r}
library(tidyverse)
library(ggplot2)
source ("./R/functions.R")
source ("./R/GWAS.R")
```

```{r}
execute = TRUE
```

Execute the GWAS
```{r message=FALSE, warning=FALSE}
disease <- "T2D"
basedir <- "./Data/"
outdir <- "./Results/Full/"

if (execute)
  GWASD (disease, basedir, outdir)
```

Read in the results of GWAS. Filter our any that have MAF < 1% and HWE p.value < 0.05

```{r}
fData <- read_tsv("./Results/T2D.tsv")
colnames(fData) <-  c("ID", "SNP","RSID","CHR", "MajA", "MinA", "MAFD","MAFC","OR", 
                      "P", "CHI2", "HWE" )
```

(1) Manhattan Plot

```{r}
manhattanPlot(fData, "ID", "P")
```

QQPlot
```{r}
p = sort(-log10(fData$P))
myqqplot(p)
```

Examining SNP rs4506565
```{r}
library(broom)
fData %>%
  filter(RSID == "rs4506565")
```

Genomic Inflation Factor
```{r}
lambda <- gif (fData)

## filter for gif
fData_corrected = fData %>% 
                  mutate(chi2_corrected = CHI2/lambda, 
                         pval_corrected = pchisq(chi2_corrected, 1, lower.tail = F))

manhattanPlot(fData_corrected, "ID", "pval_corrected")

p = sort(-log10(fData_corrected$pval_corrected))
myqqplot(p)
```

```{r}
# Before
paste(nrow(fData %>% filter(P <= p_bonferroni)), 'on',
      nrow(fData), 'total snps')
# After
paste(nrow(fData_corrected %>% filter(pval_corrected <= p_bonferroni)), 'on', 
      nrow(fData_corrected), 'total snps')
```



